MoltenVK/MoltenVK/GPUObjects/MVKDeviceMemory.mm

```diff
@@ -328,75 +177,91 @@
 					// Need to use the managed mode for images.
-					_mtlStorageMode = MTLStorageModeManaged;
-				}
-				// Nonetheless, we need a buffer to be able to map the memory at will.
-				if (!ensureMTLBuffer() ) {
-					setConfigurationResult(reportError(VK_ERROR_OUT_OF_DEVICE_MEMORY, "vkAllocateMemory(): Could not allocate a host-coherent VkDeviceMemory of size %llu bytes. The maximum memory-aligned size of a host-coherent VkDeviceMemory is %llu bytes.", _allocationSize, getMetalFeatures().maxMTLBufferSize));
+					storageMode = MTLStorageModeManaged;
+					_options = mvkMTLResourceOptions(storageMode, cpuCacheMode);
 				}
 			}
 		}
 #endif
-        for (auto& memoryBinding : ((MVKImage*)dedicatedImage)->_memoryBindings) {
-            _imageMemoryBindings.push_back(memoryBinding);
-        }
-		return;
 	}
 
-	if (dedicatedBuffer) {
-		_buffers.push_back((MVKBuffer*)dedicatedBuffer);
-	}
+	// Only allocate memory if it was not imported
+	if (_mtlBuffer)
+		return;
 
-	// If we can, create a MTLHeap. This should happen before creating the buffer, allowing us to map its contents.
-	if ( !_isDedicated ) {
-		if (!ensureMTLHeap()) {
-			setConfigurationResult(reportError(VK_ERROR_OUT_OF_DEVICE_MEMORY, "vkAllocateMemory(): Could not allocate VkDeviceMemory of size %llu bytes.", _allocationSize));
-			return;
-		}
-	}
+	MTLHeapDescriptor* heapDesc = [MTLHeapDescriptor new];
+	heapDesc.type = MTLHeapTypePlacement;
+	heapDesc.resourceOptions = _options;
+	// For now, use tracked resources. Later, we should probably default
+	// to untracked, since Vulkan uses explicit barriers anyway.
+	heapDesc.hazardTrackingMode = MTLHazardTrackingModeTracked;
+	heapDesc.size = _size;
+	_mtlHeap = [mtlDevice newHeapWithDescriptor: heapDesc];	// retained
+	[heapDesc release];
+	if (!_mtlHeap) goto fail_alloc;
+	propagateDebugName();
 
-	// If memory needs to be coherent it must reside in a MTLBuffer, since an open-ended map() must work.
-	// If memory was imported, a MTLBuffer must be created on it.
-	// Or if a MTLBuffer will be exported, ensure it exists.
-	if ((isMemoryHostCoherent() || _isHostMemImported || willExportMTLBuffer) && !ensureMTLBuffer() ) {
-		setConfigurationResult(reportError(VK_ERROR_OUT_OF_DEVICE_MEMORY, "vkAllocateMemory(): Could not allocate a host-coherent or exportable VkDeviceMemory of size %llu bytes. The maximum memory-aligned size of a host-coherent VkDeviceMemory is %llu bytes.", _allocationSize, getMetalFeatures().maxMTLBufferSize));
+	// Create a buffer that expands the whole heap
+	// to be able to map and flush as required by the application
+	if (mvkIsAnyFlagEnabled(_vkMemPropFlags, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+		_mtlBuffer = [_mtlHeap newBufferWithLength:_size options:_options];
+		if (!_mtlBuffer) goto fail_alloc;
+		[_mtlBuffer makeAliasable];
 	}
+
+fail_alloc:
+	setConfigurationResult(reportError(VK_ERROR_OUT_OF_DEVICE_MEMORY, "vkAllocateMemory(): Could not allocate VkDeviceMemory of size %llu bytes.", _size));
```

Comment:
Please don't use `goto`. Package this in a `reportAllocFail()` function, and then just call `return reportAllocFail();` in the `goto` calls above.

Also, to cover someone trying to use this with ancient OS versions, can you add a check in that function for `getMetalFeatures().placementHeaps`, and if it's false, report a slightly different message indicating that the app must be running on macOS 10.15 or iOS/tvOS 13.0. Those are also our current minimum supported versions anyway.